File: plugins/apps/qwenpaw-creator/ui/src/contracts/creator/tasks.ts
import type { SpecialistRole } from "./common";

export type SpecialistRunStatus =
  | "QUEUED"
  | "QUEUED_CAPACITY"
  | "RUNNING_MODEL"
  | "WAITING_RUNTIME"
  | "WAITING_AUTHORIZATION"
  | "SUCCEEDED"
  | "BLOCKED"
  | "FAILED"
  | "STALE"
  | "CANCELLED";

export type TaskStatus =
  | "QUEUED"
  | "RUNNING"
  | "SUCCEEDED"
  | "FAILED"
  | "CANCELLED"
  | "QUARANTINED";

export interface SpecialistRunView {
  id: string;
  role: SpecialistRole;
  displayName: string;
  status: SpecialistRunStatus;
  targetRefs: string[];
  relatedRunId?: string;
  supersedesRunId?: string;
  finalMarker?: "SUCCESS" | "BLOCKED" | "FAILED";
  finalSummaryText?: string;
  taskRefs: string[];
  transactionId?: string | null;
  createdAt?: string;
  updatedAt?: string;
  metadata: Record<string, unknown>;
}

export interface TaskView {
  id: string;
  projectId: string;
  transactionId: string | null;
  specialistRunId: string | null;
  kind:
    | "asset_ingest"
    | "asset_import"
    | "source_intelligence"
    | "image_generation"
    | "r2v_generation"
    | "ai_edit_plan"
    | "ai_edit_execute"
    | "compose";
  targetRef: string;
  status: TaskStatus;
  progress: number | null;
  completedElements?: number | null;
  totalElements?: number | null;
  resultRefs: string[];
  result?: Record<string, unknown> | null;
  error?: Record<string, unknown> | null;
  createdAt?: string;
  updatedAt?: string;
}

export interface TaskListResponse {
  items: TaskView[];
}

export interface SpecialistRunListResponse {
  items: SpecialistRunView[];
}

export interface ExecutionAuthorizationView {
  id: string;
  transactionId: string;
  specialistRunId: string;
  executionRequestId: string;
  targetRef: string;
  scope: Record<string, unknown>;
  status: "PENDING" | "APPROVED" | "DECLINED" | "EXPIRED";
  authorizationToken: string;
  provider: string;
  model: string;
  estimatedCost?: number;
  currency?: string;
  maxCandidates: number;
  createdAt: string;
}

export interface ExecutionAuthorizationDecision {
  authorizationToken: string;
}

export interface ExecutionAuthorizationApproval
  extends ExecutionAuthorizationDecision {
  provider: string;
  model: string;
  maxCost: number;
  maxCandidates: number;
}

File: plugins/apps/qwenpaw-creator/backend/schemas/tasks.py
# -*- coding: utf-8 -*-
from __future__ import annotations

from typing import Any, Literal

from pydantic import Field

from domain.enums import (
    SpecialistRole,
    SpecialistRunStatus,
    TaskKind,
    TaskStatus,
)

from .common import StrictModel


class TaskView(StrictModel):
    id: str
    project_id: str = Field(alias="projectId")
    transaction_id: str | None = Field(None, alias="transactionId")
    specialist_run_id: str | None = Field(None, alias="specialistRunId")
    kind: TaskKind
    target_ref: str = Field(alias="targetRef")
    status: TaskStatus
    progress: float | None = None
    completed_elements: int | None = Field(
        None,
        alias="completedElements",
        ge=0,
    )
    total_elements: int | None = Field(None, alias="totalElements", ge=0)
    result_refs: list[str] = Field(default_factory=list, alias="resultRefs")


class SpecialistRunView(StrictModel):
    id: str
    role: SpecialistRole
    display_name: str = Field(alias="displayName")
    status: SpecialistRunStatus
    target_refs: list[str] = Field(alias="targetRefs")
    related_run_id: str | None = Field(None, alias="relatedRunId")
    supersedes_run_id: str | None = Field(None, alias="supersedesRunId")
    final_marker: str | None = Field(None, alias="finalMarker")
    final_summary_text: str | None = Field(None, alias="finalSummaryText")
    task_refs: list[str] = Field(default_factory=list, alias="taskRefs")
    metadata: dict[str, Any] = Field(default_factory=dict)


class ExecutionAuthorizationView(StrictModel):
    id: str
    transaction_id: str = Field(alias="transactionId")
    specialist_run_id: str = Field(alias="specialistRunId")
    execution_request_id: str = Field(alias="executionRequestId")
    target_ref: str = Field(alias="targetRef")
    scope: dict[str, Any]
    status: Literal["PENDING", "APPROVED", "DECLINED", "EXPIRED"]
    authorization_token: str = Field(alias="authorizationToken")
    provider: str
    model: str
    max_candidates: int = Field(alias="maxCandidates", ge=1)
    created_at: str = Field(alias="createdAt")


class ExecutionAuthorizationPage(StrictModel):
    items: list[ExecutionAuthorizationView]
